_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,372 @@
+so that it could handle more kind of BEAM instructions.
+
+[John Högberg][john] quickly became interested in the project and
+started to act as a sounding board.  Some time later, probably in
+March, John suggested that the new JIT should translate **all** loaded
+code to native code.  That way, there would be no need to support
+context switching between the BEAM interpreter and native code, which
+would make the design simpler and eliminate the cost for context
+switches.
+
+That was a gamble, of course. After all, it could turn out that the
+native code could be too large to be practically useful or decrease
+performance because it fitted badly in the code cache.  They decided
+that it was worth taking the risk and that it would probably be
+possible to optimize the size of the code later.  (*Spoiler*: At the time
+of writing, the native code generated by the JIT is about 10 percent
+larger than interpreted BEAM code.)
+
+Another change to the design was the tooling for generated the native
+code. In Lukas's prototype, the native code template for each
+instruction was contained in text files similar to the other files
+used by the loader.  That was inflexible, so it was decided to use
+some library that could generate native code.  While some pure C
+libraries could have been used, the C++ library AsmJIT was more
+convenient in practical use than any of the C libraries.  Therefore
+the part of the loader that translates BEAM instructions to native
+code needed to be written in C++, but the rest of the runtime system
+is still pure C code and will remain so.
+
+John joined the practical work on the rejigged JIT project at the end
+of March.
+
+On April 7, 2020, John reached the "prompt beer" milestone.
+
+### Prompt Beer
+
+When the Erlang system is started, a surprisingly amount of code is
+executed before the prompt appears. On the one hand that means that
+the translation of many instructions needs to be implemented
+before it would be possible to even start the Erlang system, let alone run
+any test suites or benchmarks.
+
+On the other hand, when the prompt finally appears, it is a major
+milestone worth celebrating with some prompt beer or other appropriate
+beverage or by taking the rest of the evening off.
+
+### Maturing the new JIT
+
+The next major milestone was reached May 6 when line numbers in
+stack traces were implemented. That meant that many more test cases now
+succeeded.
+
+Soon after that, all test suites could be run successfully.  It was decided
+to develop the new JIT into a product that could be included in OTP 24.  During
+the summer and early fall, the following was done:
+
+* A major refactoring of the BEAM loader so that as much code as possible
+  could be shared between the JIT and the BEAM interpreter. (The BEAM interpreter
+  is only used on platforms that don't support the JIT.)
+
+* Implementing and polishing of important but less used features such
+  as tracing, save calls, and perf support.
```

Comment:
Maybe add a link to the `perf` tool?

Reply:
I will add a link to the Wikipedia page for perf.